weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldPortableExtensionProcessor.java

```diff
@@ -74,15 +73,13 @@ private void loadAttachments(Module module, DeploymentUnit deploymentUnit, WeldP
             // load class and register for portable extensions
             Collection<Class<?>> loadedPortableExtensions = loadExtensions(module, portableExtensionServices, Object.class);
             registerPortableExtensions(deploymentUnit, extensions, loadedPortableExtensions);
-            // load class and register for portable extensions
-            // if there is at least one, add a portable extension processing them
-            List<Class<? extends BuildCompatibleExtension>> loadedBuildCompatExtensions =
-                    loadExtensions(module, buildCompatibleExtensionServices, BuildCompatibleExtension.class);
-            if (!loadedBuildCompatExtensions.isEmpty()) {
-                Extension extension = new LiteExtensionTranslator(loadedBuildCompatExtensions, module.getClassLoader());
-                // NOTE: I chose to register it under the same dep. unit as other extensions, not sure if this is correct
-                extensions.registerExtensionInstance(extension, deploymentUnit);
+            // load class and register for discovered build compatible extensions
+            for (Class<? extends BuildCompatibleExtension> extensionClass : loadExtensions(module, buildCompatibleExtensionServices, BuildCompatibleExtension.class)) {
+                extensions.registerBuildCompatibleExtension(extensionClass);
             }
+            // prepare metadata for later registration of LiteExtensionTranslator
+            // NOTE: I chose to register it under the same dep. unit as other extensions, not sure if this is correct
+            extensions.prepareLiteExtensionTranslator(module.getClassLoader(), deploymentUnit);
```

Comment:
I might be understanding this incorrectly, but it seems that WeldPortableExtensions is scoped to the top level deployment unit, while this method executes for both top level deployments and subdeployments. So I think subdeployments will end up overwriting each other.

Reply:
Hm, that would mean the previous code was also incorrect.
Maybe I could use the CL and DU taken from the `WeldDeploymentProcessor`. Does the top level deployment always have visibility to all subdeployments?

Reply:
I've done the above and re-tested the scenario we talked through with Emmanuel and that works fine as do other tests with BCE within CDI TCKs. I'll look into writing the test now and will probably ping you either today or tomorrow.